pkg/snapshot/restorer/restorer_test.go

```diff
@@ -654,3 +894,89 @@ func corruptEtcdDir() error {
 	}
 	return os.RemoveAll(etcdDir)
 }
+
+//takeValidV1Snaps saves valid snaps in the v1 prefix dir of snapstore so that restorer could restore from them
+//TODO: Consider removing when backward compatibility no longer needed
+func takeValidV1Snaps(logger *logrus.Entry, container string, deltaSnapshotPeriod time.Duration, endpoints []string, stopCh <-chan struct{}, startWithFullSnapshot bool, compressionConfig *compressor.CompressionConfig) error {
+	//Here we run the snapshotter to take snapshots. The snapshotter by default stores the snaps in the v2 directory.
+	//We then move those snaps into the v1 dir under a 'Backup-xxxxxx' dir
+
+	//Snapshots for the v2 dir
+	//Add data into etcd
+	resp := &utils.EtcdDataPopulationResponse{}
+	utils.PopulateEtcd(testCtx, logger, endpoints, 0, keyTo, resp)
+	Expect(resp.Err).ShouldNot(HaveOccurred())
+	//Take a full snapshot.
+	ctx, cancel := context.WithTimeout(testCtx, time.Duration(2*time.Second))
+	snapstoreConfig := brtypes.SnapstoreConfig{Container: snapstoreDir, Provider: "Local", Prefix: "v2"}
+	err = utils.RunSnapshotter(logger, snapstoreConfig, deltaSnapshotPeriod, endpoints, ctx.Done(), true, compressionConfig)
+	Expect(err).ShouldNot(HaveOccurred())
+	cancel()
+
+	//Add data into etcd
+	utils.PopulateEtcd(testCtx, logger, endpoints, 0, keyTo, resp)
+	Expect(resp.Err).ShouldNot(HaveOccurred())
+	//Take delta snapshot.
+	ctx, cancel = context.WithTimeout(testCtx, time.Duration(2*time.Second))
+	snapstoreConfig = brtypes.SnapstoreConfig{Container: snapstoreDir, Provider: "Local", Prefix: "v2"}
+	err = utils.RunSnapshotter(logger, snapstoreConfig, deltaSnapshotPeriod, endpoints, ctx.Done(), false, compressionConfig)
+	Expect(err).ShouldNot(HaveOccurred())
+	cancel()
+
+	//Add data to etcd
+	utils.PopulateEtcd(testCtx, logger, endpoints, 0, keyTo, resp)
+	Expect(resp.Err).ShouldNot(HaveOccurred())
+	//Take delta snapshot.
+	ctx, cancel = context.WithTimeout(testCtx, time.Duration(2*time.Second))
+	snapstoreConfig = brtypes.SnapstoreConfig{Container: snapstoreDir, Provider: "Local", Prefix: "v2"}
+	err = utils.RunSnapshotter(logger, snapstoreConfig, deltaSnapshotPeriod, endpoints, ctx.Done(), false, compressionConfig)
+	Expect(err).ShouldNot(HaveOccurred())
+	cancel()
+
+	//Move snaps from v2 dir to a v1 dir
+	//Create v1/Backup-xxxxxx dir
+	var curTime = time.Now().Unix()
+	err = os.MkdirAll(path.Join(path.Join(snapstoreDir, "v1"), fmt.Sprintf("Backup-%d", curTime)), 0755)
+	Expect(err).ShouldNot(HaveOccurred())
+	//Move contents from v2 to v1/Backup-xxxxxx
+	files, err := ioutil.ReadDir(path.Join(snapstoreDir, "v2"))
+	Expect(err).ShouldNot(HaveOccurred())
+	oldPath := path.Join(snapstoreDir, "v2")
+	newPath := path.Join(path.Join(snapstoreDir, "v1"), fmt.Sprintf("Backup-%d", curTime))
+	for _, f := range files {
+		err = os.Rename(path.Join(oldPath, f.Name()), path.Join(newPath, f.Name()))
+		Expect(err).ShouldNot(HaveOccurred())
+	}
```

Comment:
I like this approach of taking snapshots into `v2` and then moving to `v1`. This is a good way to test backward compatibility for restoration when we have stopped support for uploading to v1 👍 

The code above moves everything under `v2` to `v1/Backup-XXX`, including the snapshots taken before this function is called which would include the base full snapshot. This has the good effect that restoration from base snapshot from the pure `v1` structure is tested.

To test the mixed case, could you please support the case where the first few snapshots (full snapshot and some incremental snapshots) are in the `v1/Backup-XXX` and some more incremental snapshots (at least one) are in `v2`.  This could possibly be controlled via a parameter to the function.

Reply:
Added a test case to support this scenario